### PR TITLE
Fix #5197, handling of algebraic universes

### PR DESCRIPTION
**Pull request by mattam82**
They were allowed to spli in terms in some cases. We now ensure that if
an evar is defined as e.g. fun x => Type@{foo}, foo is properly
refreshed to be non-algebraic as it could otherwise appear in the term
and break the invariant.

Also cleanup the implementation of refresh_universes to avoid using a
mutable reference and simply rely on the Constr.map smartmap idiom
instead.

This might have compatibility issues, e.g. in HoTT where maybe more
non-algebraic proxy universes could be generated, we'll see. If so, I'll change
the milestone to 8.8.

For the bug report proper, there is a lack of bidirectional
type-checking that makes the initial definition fail (there's a
non-canonical choice of dependency if we don't consider the typing
constraint). With the Program bidir hint it passes.

### CODE REVIEW

**Review by ppedrot — COMMENTED** [review]
I'm afraid this kind of change is beyond my comprehension of universe unification, I'll have to admit being unable to review it...

**Comment by mattam82** [comment]
That's ok, trust me on this one :) I'm curious about the performance impact of this if any.

**Comment by maximedenes** [comment]
I'll have a look and ask for more comments if I need some.

**Comment by ejgallego** [comment]
> That's ok, trust me on this one :) I'm curious about the performance impact of this if any.

I can launch a test on my system, what are some interesting contribs to try?

**Comment by mattam82** [comment]
HoTT, once I figure out how to fix it :)

**Comment by ejgallego** [comment]
> HoTT, once I figure out how to fix it :)

Ok, adding a recipe for HoTT.

**Comment by mattam82** [comment]
I fixed the patch, now HoTT passes locally, this should be good to go.

**Comment by mattam82** [comment]
It passed on my travis:

 https://travis-ci.org/mattam82/coq/builds/308116526?utm_source=email&utm_medium=notification

Maybe @herbelin or @SkySkimmer can give a review. It's mainly about how do we refresh algebraic universes.

**Comment by maximedenes** [comment]
Sorry for the delay, I'll have a look in the afternoon, but other reviews are indeed welcome.

File: pretyping/evarsolve.ml [hunk 1]
```diff
@@ -73,25 +72,33 @@ let refresh_universes ?(status=univ_rigid) ?(onlyalg=false) ?(refreshset=false)
...
       let evi = Evd.find !evdref ev in
-      let ty' = refresh ~onlyalg univ_flexible ~direction:true (EConstr.of_constr evi.evar_concl) in
-	if !modified then 
-	  evdref := Evd.add !evdref ev {evi with evar_concl = EConstr.Unsafe.to_constr ty'}
-	else ()
-    | _ -> EConstr.iter !evdref (refresh_term_evars onevars false) t
+      let ty = EConstr.of_constr evi.evar_concl in
+      let ty' = refresh ~onlyalg univ_flexible ~direction:true ty in
+      if ty == ty' then t
+      else (evdref := Evd.add !evdref ev {evi with evar_concl = EConstr.to_constr !evdref ty'}; t)
```
Comment (SkySkimmer) [inline pretyping/evarsolve.ml]: `Evd.add` has a comment:
```ocaml
(** [add sigma ev info] adds [ev] with evar info [info] in sigma.
    Precondition: ev must not preexist in [sigma]. *)
```
Is it outdated or what?

**Review by SkySkimmer — COMMENTED** [review]
I don't really understand how refreshing works but it seems like it shouldn't regress at least.

File: pretyping/evarsolve.ml (hunk 1 above)
Comment (ppedrot) [reply pretyping/evarsolve.ml]: That's a good practice comment, but a few people break this invariant, *cough* ssr *cough*. We should provide a function for evar updates.

**Comment by maximedenes** [comment]
> Ok, adding a recipe for HoTT.

@ejgallego did you get any number for the performance of this patch?

File: pretyping/pretyping.ml [hunk 2]
```diff
@@ -518,17 +518,24 @@ let pretype_ref ?loc evdref env ref us =
     let ty = unsafe_type_of env.ExtraEnv.env evd c in
       make_judge c ty
 
-let judge_of_Type ?loc evd s =
-  let evd, s = interp_universe ?loc evd s in
+let judge_of_Type ?loc env evd s =
```
Comment (maximedenes) [inline pretyping/pretyping.ml]: Is this directly related to the rest of the changes?

**Review by maximedenes — DISMISSED** [review]

**Comment by ejgallego** [comment]
> @ejgallego did you get any number for the performance of this patch?

I may need a couple of days unfortunately so maybe a ci.inria job would be faster.

File: pretyping/evarsolve.ml [hunk 3]
```diff
@@ -73,25 +72,33 @@ let refresh_universes ?(status=univ_rigid) ?(onlyalg=false) ?(refreshset=false)
...
-	refresh_polymorphic_positions args pos
+        refresh_polymorphic_positions args pos; t
     | App (f, args) when top && isEvar !evdref f -> 
-      refresh_term_evars true false f; 
-      Array.iter (refresh_term_evars onevars false) args
+       let f' = refresh_term_evars true false f in
+       let args' = Array.map (refresh_term_evars onevars false) args in
+       if f' == f && args' == args then t
+       else mkApp (f', args')
     | Evar (ev, a) when onevars ->
```
Comment (maximedenes) [inline pretyping/evarsolve.ml]: What happens if the `Evar` has no instantiation, but candidates containing algebraic universes? Can that happen?

File: pretyping/evarsolve.ml [hunk 4]
```diff
@@ -73,25 +72,33 @@ let refresh_universes ?(status=univ_rigid) ?(onlyalg=false) ?(refreshset=false)
...
-	  evdref := Evd.add !evdref ev {evi with evar_concl = EConstr.Unsafe.to_constr ty'}
-	else ()
-    | _ -> EConstr.iter !evdref (refresh_term_evars onevars false) t
+      let ty = EConstr.of_constr evi.evar_concl in
+      let ty' = refresh ~onlyalg univ_flexible ~direction:true ty in
+      if ty == ty' then t
+      else (evdref := Evd.add !evdref ev {evi with evar_concl = EConstr.to_constr !evdref ty'}; t)
+    | Sort s ->
+       (match ESorts.kind !evdref s with
+        | Type u when not (Univ.Universe.is_levels u) ->
```
Comment (maximedenes) [inline pretyping/evarsolve.ml]: I don't understand this. If I call `refresh_universes` with `~onlyalg:false`, I'd expect all universes to be refreshed here, not just algebraics. Am I missing something? I would expect instead to call `refresh` on the sort.

File: pretyping/evarsolve.ml (hunk 1 above)
Comment (maximedenes) [reply pretyping/evarsolve.ml]: @gares asks if this doesn't break the idea behind the monotonic interface, in the sense that you have `t` and `t'` in `sigma`, you refresh `t`, and get `t1` and `sigma1'. Now if you typecheck `t'` in `sigma1`, you crash.

What do you mean by an evar update function, in the context of such monotonicity invariants?

File: pretyping/evarsolve.ml [hunk 5]
```diff
@@ -40,23 +40,22 @@ let get_polymorphic_positions sigma f =
...
     match EConstr.kind !evdref t with
     | Sort s ->
       begin match ESorts.kind !evdref s with
       | Type u ->
-       (match Univ.universe_level u with
+         (* TODO: check if max(l,u) is not ok as well *)
+        (match Univ.universe_level u with
 	| None -> refresh_sort status ~direction s
 	| Some l ->
 	   (match Evd.universe_rigidity evd l with
```
[the remark this answers is not in the document]
Comment (mattam82) [reply pretyping/evarsolve.ml]: Right

File: pretyping/evarsolve.ml [hunk 6]
```diff
@@ -73,25 +72,33 @@ let refresh_universes ?(status=univ_rigid) ?(onlyalg=false) ?(refreshset=false)
...
       | _ -> t
       end
-    | Prod (na,u,v) -> 
-      mkProd (na, u, refresh ~onlyalg status ~direction v)
+    | Prod (na,u,v) ->
+       let v' = refresh ~onlyalg status ~direction v in
+       if v' == v then t else mkProd (na, u, v')
     | _ -> t
   (** Refresh the types of evars under template polymorphic references *)
   and refresh_term_evars onevars top t =
```
[the remark this answers is not in the document]
Comment (mattam82) [reply pretyping/evarsolve.ml]: Indeed.

[the remark this answers is not in the document]
Comment (mattam82) [reply pretyping/evarsolve.ml]: Yes

File: pretyping/pretyping.ml (hunk 2 above)
Comment (mattam82) [reply pretyping/pretyping.ml]: It could be an independent PR. I just spotted that it was potentially breaking the no-algebraic  in term position invariant.

File: pretyping/evarsolve.ml (hunk 3 above)
Comment (mattam82) [reply pretyping/evarsolve.ml]: It can happen but it is fine. We are only interested to ensure that the declared type of the evar is refreshed.

Comment (maximedenes) [reply pretyping/evarsolve.ml]: @mattam82 answered on gitter that if resolution by a candidate uses `evar_define`, it should be fine.

Comment (maximedenes) [reply pretyping/evarsolve.ml]: Oops! Interesting race condition :)

**Comment by Zimmi48** [comment]
I moved the milestone to 8.8 following the compatibility problem on HoTT but if that is solved now maybe this PR has become once again a candidate for 8.7.1. WDYT?

**Comment by maximedenes** [comment]
@mattam82 please update the milestone to reflect your current schedule on this.

**Comment by herbelin** [comment]
Sorry, I missed the request for a review. Naive question: is it a case where a refreshing could be done at evar instantiation (since, originally, `refresh_universes` was for arities only iirc, and along this rationale, substitution of algebraic universes in subterms would have to be controlled at substitution time?), or do I miss the source of this algebraic universe to refresh? 
In any  case, I don't see something critical here, so it seems ok to me as it is now.

**Comment by maximedenes** [comment]
@mattam82 ping

**Comment by mattam82** [comment]
Changed milestone to 8.8.0, it should only be a bugfix. Hopefully can be finished by 8.8.0 but I've more pressing matters today I think.

**Comment by maximedenes** [comment]
> Changed milestone to 8.8.0, it should only be a bugfix. Hopefully can be finished by 8.8.0 but I've more pressing matters today I think.

@mattam82 Any update?

**Review by herbelin — COMMENTED** [review]
Formally validating the review button: see above for my comments.

**Comment by maximedenes** [comment]
@mattam82 ping. Let me know if we should expect this for 8.8.0.

**Comment by maximedenes** [comment]
@mattam82 I got no news, so delaying to 8.8.1. If you plan to review it today, please let me know ASAP. Thanks!

**Comment by Zimmi48** [comment]
Looks like this won't be ready in time for 8.8.1 either.

File: pretyping/evarsolve.ml (hunk 1 above)
Comment (mattam82) [reply pretyping/evarsolve.ml]: Refreshing ensures that any solution of the initial evars is typable with the refreshed type if it was typable with the non-refreshed one. Replaced this with Evd.downcast which does precisely this.

File: pretyping/evarsolve.ml (hunk 4 above)
Comment (mattam82) [reply pretyping/evarsolve.ml]: ~onlyalg:true means algebraics are refreshed _and_ local universe levels which are algebraic (as declared in the evd) gets their status changed to non-algebraic. This ensures that the refreshed type can be used as a term (it won't contain algebraics). ~onlyalg:false means all algebraics and non-local universes. In that case we want to ensure (for template polymorphism) that fresh universes are generated replacing the global universes, to avoid capturing them, which would break the feature. Refreshing does not need to be done on non-algebraic local universes.

File: pretyping/pretyping.ml (hunk 2 above)
Comment (mattam82) [reply pretyping/pretyping.ml]: This closes recent bug reports I believe, where Type@{i+1} is put in the wrong place.

**Comment by mattam82** [comment]
Fixed and rebase

**Comment by Zimmi48** [comment]
@mattam82 It looks like it doesn't compile.

**Comment by Zimmi48** [comment]
Feel free to re-add the 8.8.1 milestone if it's still open when this is reviewed and ready to merge.

**Comment by ppedrot** [comment]
Let me take ownership of this PR, otherwise we will never converge. @mattam82 can you rebase on a recent master?

**Comment by mattam82** [comment]
Ok, I’ll do it ASAP.

**Comment by mattam82** [comment]
@ppedrot done

File: pretyping/evarsolve.ml [hunk 7]
```diff
@@ -73,25 +72,33 @@ let refresh_universes ?(status=univ_rigid) ?(onlyalg=false) ?(refreshset=false)
...
     match EConstr.kind !evdref t with
     | App (f, args) when is_template_polymorphic env !evdref f ->
       let pos = get_polymorphic_positions !evdref f in
-	refresh_polymorphic_positions args pos
+        refresh_polymorphic_positions args pos; t
     | App (f, args) when top && isEvar !evdref f -> 
-      refresh_term_evars true false f; 
-      Array.iter (refresh_term_evars onevars false) args
+       let f' = refresh_term_evars true false f in
```
[the remark this answers is not in the document]
Comment (mattam82) [reply pretyping/evarsolve.ml]: Yes, if the evar is at the toplevel of the term and we want a fresh type for that term we force refreshing that specific evar.

File: pretyping/evarsolve.ml (hunk 6 above)
[the remark this answers is not in the document]
Comment (mattam82) [reply pretyping/evarsolve.ml]: Done

**Comment by ppedrot** [comment]
@mattam82 Is equations broken due to this PR or was it because of the merge delay? Otherwise, is this mergeable? In particular, @maximedenes are you happy with the current state?

**Comment by ppedrot** [comment]
This looks OK to me. @maximedenes I plan on merging this PR tomorrow, voice up if unhappy.

**Comment by vbgl** [comment]
This won’t pass the test-suite: `5197` is not a valid module name.

**Comment by ppedrot** [comment]
Hum, didn't it pass the CI already? Or was there a race?

**Comment by vbgl** [comment]
That’s a race… #8679 has just got merged.

**Comment by ejgallego** [comment]
> Hum, didn't it pass the CI already? Or was there a race?

There was a race with the test-suite changes. An easy one to fix fortunately.

**Comment by ppedrot** [comment]
OK, can you open a PR to fix this swiftly?

**Comment by vbgl** [comment]
Taking care of it.
